src/SFA.DAS.FAT.Application/Shortlist/Services/ShortlistService.cs

```diff
@@ -6,48 +6,68 @@
+
+    public ShortlistService(IApiClient apiClient, IOptions<FindApprenticeshipTrainingApi> config, ISessionService sessionService)
     {
-        private readonly IApiClient _apiClient;
-        private readonly FindApprenticeshipTrainingApi _configValue;
+        _apiClient = apiClient;
+        _sessionService = sessionService;
+        _configValue = config.Value;
+    }
 
-        public ShortlistService(IApiClient apiClient, IOptions<FindApprenticeshipTrainingApi> config)
+    public async Task<int> GetShortlistsCountForUser(Guid shortlistUserId)
+    {
+        var shortlistCount = _sessionService.Get<ShortlistsCount>();
+        if (shortlistCount == null)
         {
-            _apiClient = apiClient;
-            _configValue = config.Value;
+            var apiRequest = new GetShortlistsCountForUserRequest(_configValue.BaseUrl, shortlistUserId);
+            shortlistCount = await _apiClient.Get<ShortlistsCount>(apiRequest);
+            _sessionService.Set(shortlistCount);
         }
+        return shortlistCount.Count;
+    }
 
-        public async Task<ShortlistForUser> GetShortlistForUser(Guid shortlistUserId)
-        {
-            var apiRequest = new GetShortlistForUserApiRequest(_configValue.BaseUrl, shortlistUserId);
+    public async Task<ShortlistForUser> GetShortlistForUser(Guid shortlistUserId)
+    {
+        var apiRequest = new GetShortlistForUserApiRequest(_configValue.BaseUrl, shortlistUserId);
 
-            var apiResponse = await _apiClient.Get<ShortlistForUser>(apiRequest);
+        var apiResponse = await _apiClient.Get<ShortlistForUser>(apiRequest);
 
-            return apiResponse;
-        }
+        return apiResponse;
+    }
+
+    public async Task DeleteShortlistItemForUser(Guid id)
+    {
+        await _apiClient.Delete(new DeleteShortlistForUserRequest(_configValue.BaseUrl, id));
 
-        public async Task DeleteShortlistItemForUser(Guid id, Guid shortlistUserId)
+        var shortlistCount = _sessionService.Get<ShortlistsCount>();
+        if (shortlistCount != null)
         {
-            await _apiClient.Delete(new DeleteShortlistForUserRequest(_configValue.BaseUrl, id, shortlistUserId));
+            shortlistCount.Count--;
+            _sessionService.Set(shortlistCount);
         }
+    }
 
-        public async Task<Guid> CreateShortlistItemForUser(Guid shortlistUserId, int ukprn, int trainingCode, double? lat, double? lon, string locationDescription)
+    public async Task<Guid> CreateShortlistItemForUser(PostShortlistForUserRequest request)
+    {
+        var response = await _apiClient.Post<CreateShortlistForUserResponse, PostShortlistForUserRequest>(new CreateShortlistForUserRequest(_configValue.BaseUrl) { Data = request });
+
+        if (response.IsCreated)
         {
-            var request = new PostShortlistForUserRequest
+            var shortlistCount = _sessionService.Get<ShortlistsCount>();
+            if (shortlistCount != null)
```

Comment:
It's a bit difficult to work out just by looking at the code, but is it possible the shortlistcount will be null if this is the first 'add' against the user?  in which case the 'if' condition shouldn't be tested against?

Reply:
We only want to update session if we can retrieve it, if there is nothing in the session, the view component will pull the count from the api and create one, so no need to create here. I know it is not obvious, i could have done it here, but the component is self correcting.